internal/operator-controller/authorization/rbac.go

```diff
@@ -0,0 +1,576 @@
+		ruleResolver: newRBACRulesResolver(cl),
+		restMapper:   cl.RESTMapper(),
+	}
+}
+
+// PreAuthorize validates whether the current user/request satisfies the necessary permissions
+// as defined by the RBAC policy. It examines the user’s roles, resource identifiers, and
+// the intended action to determine if the operation is allowed.
+//
+// Return Value:
+//   - nil: indicates that the authorization check passed and the operation is permitted.
+//   - non-nil error: indicates that an error occurred during the permission evaluation process
+//     (for example, a failure decoding the manifest or other internal issues). If the evaluation
+//     completes successfully but identifies missing rules, then a nil error is returned along with
+//     the list (or slice) of missing rules. Note that in some cases the error may encapsulate multiple
+//     evaluation failures
+func (a *rbacPreAuthorizer) PreAuthorize(ctx context.Context, ext *ocv1.ClusterExtension, manifestReader io.Reader) ([]ScopedPolicyRules, error) {
+	var allMissingPolicyRules = []ScopedPolicyRules{}
+	dm, err := a.decodeManifest(manifestReader)
+	if err != nil {
+		return nil, err
+	}
+	manifestManager := &user.DefaultInfo{Name: fmt.Sprintf("system:serviceaccount:%s:%s", ext.Spec.Namespace, ext.Spec.ServiceAccount.Name)}
+	attributesRecords := dm.asAuthorizationAttributesRecordsForUser(manifestManager, ext)
+
+	var preAuthEvaluationErrors []error
+	missingRules, err := a.authorizeAttributesRecords(ctx, attributesRecords)
+	if err != nil {
+		preAuthEvaluationErrors = append(preAuthEvaluationErrors, err)
+	}
+
+	ec := escalationChecker{
+		authorizer:        a.authorizer,
+		ruleResolver:      a.ruleResolver,
+		extraClusterRoles: dm.clusterRoles,
+		extraRoles:        dm.roles,
+	}
+
+	for _, obj := range dm.rbacObjects() {
+		if err := ec.checkEscalation(ctx, manifestManager, obj); err != nil {
+			missingEscalationRules, err := parseEscalationErrorForMissingRules(err)
+			missingRules[obj.GetNamespace()] = append(missingRules[obj.GetNamespace()], missingEscalationRules...)
+			preAuthEvaluationErrors = append(preAuthEvaluationErrors, err)
+		}
+	}
+
+	for ns, nsMissingRules := range missingRules {
+		// NOTE: Although CompactRules is defined to return an error, its current implementation
+		// never produces a non-nil error. This is because all operations within the function are
+		// designed to succeed under current conditions. In the future, if more complex rule validations
+		// are introduced, this behavior may change and proper error handling will be required.
+		if compactMissingRules, err := validation.CompactRules(nsMissingRules); err == nil {
+			missingRules[ns] = compactMissingRules
+		}
+
+		missingRulesWithDeduplicatedVerbs := []rbacv1.PolicyRule{}
+		for _, rule := range missingRules[ns] {
+			verbSet := sets.New[string](rule.Verbs...)
+			if verbSet.Has("*") {
+				rule.Verbs = []string{"*"}
+			} else {
+				rule.Verbs = sets.List(verbSet)
+			}
```

Comment:
suggestion: move this to a function.

Should we also sort `rule.Verbs = sets.List(verbSet)`? I wonder if the the order here could affect the sort order in line 126? (and if there would be any follow on affects of that...)

Reply:
`sets.List` sorts, so all good.

Reply:
yep it is sorted slice so resolving this one